src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java

```diff
@@ -305,6 +313,56 @@ private Collection<InvoiceItem> toInvoiceItems(final UUID invoiceId,
+            return taxItem;
+        }
+
+        final String taxItemDetails = createTaxItemDetails(taxRate);
+        if (taxItemDetails == null) {
+            return taxItem;
+        }
+
+        return new PluginInvoiceItem(new Builder<>()
+                                             .withId(taxItem.getId())
+                                             .withInvoiceItemType(taxItem.getInvoiceItemType())
+                                             .withInvoiceId(taxItem.getInvoiceId())
+                                             .withAccountId(taxItem.getAccountId())
+                                             .withChildAccountId(taxItem.getChildAccountId())
+                                             .withStartDate(taxItem.getStartDate())
+                                             .withEndDate(taxItem.getEndDate())
+                                             .withAmount(taxItem.getAmount())
+                                             .withCurrency(taxItem.getCurrency())
+                                             .withDescription(taxItem.getDescription())
+                                             .withSubscriptionId(taxItem.getSubscriptionId())
+                                             .withBundleId(taxItem.getBundleId())
+                                             .withCatalogEffectiveDate(taxItem.getCatalogEffectiveDate())
+                                             .withProductName(taxItem.getProductName())
+                                             .withPrettyProductName(taxItem.getPrettyProductName())
+                                             .withPlanName(taxItem.getPlanName())
+                                             .withPrettyPlanName(taxItem.getPrettyPlanName())
+                                             .withPhaseName(taxItem.getPhaseName())
+                                             .withPrettyPhaseName(taxItem.getPrettyPhaseName())
+                                             .withRate(taxItem.getRate())
+                                             .withLinkedItemId(taxItem.getLinkedItemId())
+                                             .withUsageName(taxItem.getUsageName())
+                                             .withPrettyUsageName(taxItem.getPrettyUsageName())
+                                             .withQuantity(taxItem.getQuantity())
+                                             .withItemDetails(taxItemDetails)
+                                             .withCreatedDate(taxItem.getCreatedDate())
+                                             .withUpdatedDate(taxItem.getUpdatedDate())
+                                             .validate().build());
+    }
+
+    private String createTaxItemDetails(@Nonnull final Double effectiveRate) {
+        try {
+            return objectMapper.writeValueAsString(ImmutableMap.of("taxRate", String.valueOf(effectiveRate)));
+        } catch (JsonProcessingException ignored) {
+            return null;
+        }
```

Comment:
also, we whouldn't have a tax-item without eefective tax rate info. If we got null or 0 values from vertex we shouldn't create the tax item at all as I mentioned in my other comment.

Reply:
I wanted to not to change existing behaviour, imo it's safe implementation to not miss tax if effectiveTaxRate is not presented by vertex by some reason. We return not derived basic taxItem created by existing method.

```
final String taxItemDetails = createTaxItemDetails(taxRate);
        if (taxItemDetails == null) {
            return taxItem;
        }
```

Reply:
discussed